File: src/main/scala/com/exasol/cloudetl/parallelism/AbstractUdfCountCalculator.scala
package com.exasol.cloudetl.parallelism

import java.math.BigInteger

import com.exasol.ExaMetadata

/**
 * An implementation of {@link UdfCountCalculator} that uses fixed memory value for an UDF.
 */
final case class FixedUdfCountCalculator(metadata: ExaMetadata)
    extends AbstractUdfCountCalculator(metadata, 500 * 1000000)

/**
 * An implementation of {@link UdfCountCalculator} that uses dynamic memory value for an UDF.
 */
final case class MemoryUdfCountCalculator(metadata: ExaMetadata, memory: Long)
    extends AbstractUdfCountCalculator(metadata, memory)

/**
 * An abstract class for {@link UdfCountCalculator} that contains common methods.
 */
abstract class AbstractUdfCountCalculator(metadata: ExaMetadata, memory: Long) extends UdfCountCalculator {

  override def getNumberOfNodes(): Int =
    metadata.getNodeCount().toInt

  override def getNumberOfCoresLimitedByMemoryPerNode(): Int = {
    val memoryLimit = metadata.getMemoryLimit()
    val coresPerNode = memoryLimit.divide(BigInteger.valueOf(memory)).intValue()
    logger.info(
      s"Calculated cores limited by memory ($memoryLimit MB memory per node / $memory MB UDF memory = $coresPerNode)."
    )
    coresPerNode
  }

}

File: src/main/java/com/exasol/cloudetl/parallelism/AbstractUdfCountCalculator.java
package com.exasol.cloudetl.parallelism;

import java.math.BigInteger;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.exasol.ExaMetadata;

/** Common implementation for UDF count calculators. */
public abstract class AbstractUdfCountCalculator implements UdfCountCalculator {
    private static final Logger LOGGER = LoggerFactory.getLogger(AbstractUdfCountCalculator.class);

    private final ExaMetadata metadata;
    private final long memory;

    /** Create a calculator. */
    protected AbstractUdfCountCalculator(final ExaMetadata metadata, final long memory) {
        this.metadata = metadata;
        this.memory = memory;
    }

    @Override
    public int getNumberOfNodes() {
        return (int) this.metadata.getNodeCount();
    }

    @Override
    public int getNumberOfCoresLimitedByMemoryPerNode() {
        final BigInteger memoryLimit = this.metadata.getMemoryLimit();
        final int coresPerNode = memoryLimit.divide(BigInteger.valueOf(this.memory)).intValue();
        LOGGER.info("Calculated cores limited by memory ({} MB memory per node / {} MB UDF memory = {}).", memoryLimit,
                this.memory, coresPerNode);
        return coresPerNode;
    }
}
